Sources/NIOCore/NIOSplitLinesMessageDecoder.swift

```diff
@@ -0,0 +1,332 @@
+                // Safe to force unwrap. `buffer.readableBytes` is `0` in the worst case.
+                let slice = buffer.readSlice(length: buffer.readableBytes)!
+
+                return (slice, nil)
+            }
+
+            // Safe to force unwrap. We just found a separator somewhere in the buffer.
+            let slice = buffer.readSlice(length: separatorIndex - buffer.readerIndex)!
+
+            if self.omittingEmptySubsequences,
+                slice.readableBytes == 0
+            {
+                // Mark the separator itself as read
+                buffer._moveReaderIndex(forwardBy: 1)
+                continue
+            }
+
+            // Read the separator itself
+            // Safe to force unwrap. We just found a separator somewhere in the buffer.
+            let separator = buffer.readInteger(as: UInt8.self)!
+
+            return (slice, separator)
+        }
+    }
+
+    /// Decode the next message separated by the provided separator.
+    /// To be used when we're still receiving data.
+    @inlinable
+    mutating func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
+        try self.decode(buffer: &buffer, hasReceivedLastChunk: false)?.buffer
+    }
+
+    /// Decode the next message separated by the provided separator.
+    /// To be used when the last chunk of data has been received.
+    @inlinable
+    mutating func decodeLast(buffer: inout ByteBuffer, seenEOF: Bool) throws -> InboundOut? {
+        try self.decode(buffer: &buffer, hasReceivedLastChunk: true)?.buffer
+    }
+}
+
+@available(*, unavailable)
+extension SplitMessageDecoder: Sendable {}
+
+// MARK: - NIOSplitLinesMessageDecoder
+
+/// A decoder which splits the data into subsequences that are separated by a line break.
+///
+/// Use `AsyncSequence/splitLines(omittingEmptySubsequences:maximumBufferSize:)` to create a
+/// `NIODecodedAsyncSequence` that uses this decoder.
+///
+/// The following Characters are considered line breaks, similar to
+/// standard library's `String.split(whereSeparator: \.isNewline)`:
+/// - "\n" (U+000A): LINE FEED (LF)
+/// - U+000B: LINE TABULATION (VT)
+/// - U+000C: FORM FEED (FF)
+/// - "\r" (U+000D): CARRIAGE RETURN (CR)
+/// - "\r\n" (U+000D U+000A): CR-LF
+///
+/// The following Characters are NOT considered line breaks, unlike in
+/// standard library's `String.split(whereSeparator: \.isNewline)`:
+/// - U+0085: NEXT LINE (NEL)
+/// - U+2028: LINE SEPARATOR
+/// - U+2029: PARAGRAPH SEPARATOR
+///
+/// This is because these characters would require unicode and data-encoding awareness, which
+/// are outside swift-nio's scope.
+///
+/// Usage:
+/// ```swift
+/// let baseSequence = MyAsyncSequence<ByteBuffer>(...)
+/// let splitLinesSequence = baseSequence.splitLines()
+///
+/// for try await buffer in splitLinesSequence {
+///     print("Split by line breaks!\n", buffer.hexDump(format: .detailed))
+/// }
+/// ```
+public struct NIOSplitLinesMessageDecoder: NIOSingleStepByteToMessageDecoder {
```

Comment:
Should this name include `Slice` like the extension since it emits `ByteBuffer`s? It's not hard to imagine there being a `String` equivalent of this decoder.